### PR TITLE
remove read-only statement from transaction auto retry

### PR DESCRIPTION
**Pull request by wentaoxu**
1. remove explain statement from retry
1. remove all select statement from retry except It has setvar functions
1. remove execute statement from retry if it is read-only statement
@shenli @coocood

### CODE REVIEW

File: executor/compiler.go [hunk 1]
```diff
@@ -50,12 +50,18 @@ func (c *Compiler) Compile(ctx context.Context, stmtNode ast.StmtNode) (*ExecStm
 		return nil, errors.Trace(err)
 	}
 
+	ReadOnlyCheckStmt := stmtNode
```
Comment (shenli) [inline executor/compiler.go]: For local variables, we usually do not need to capitalize the name.

File: executor/read_only_checker.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	readOnly bool
+}
+
+// Enter implements Visitor interface.
+func (checker *readOnlyChecker) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
+	switch node := in.(type) {
+	case *ast.VariableExpr:
+		// like func rewriteVariable(), this stands for SetVar.
+		if !node.IsSystem && node.Value != nil {
+			checker.readOnly = false
```
Comment (shenli) [inline executor/read_only_checker.go]: Should we return true for skipChildren?

File: executor/compiler.go (hunk 1 above)
Comment (wentaoxu) [reply executor/compiler.go]: ok

File: executor/read_only_checker.go (hunk 2 above)
Comment (wentaoxu) [reply executor/read_only_checker.go]: yes, it's right

**Comment by coocood** [comment]
For prepared statement, `Compile` is not called, they use `CompileExecutePreparedStmt`.

**Comment by wentaoxu** [comment]
@coocood yes, I lost this function.

File: executor/read_only_checker.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// limitations under the License.
+
+package executor
+
+import (
+	"github.com/pingcap/tidb/ast"
+)
+
+// IsReadOnly checks whether the input ast is readOnly.
+func IsReadOnly(node ast.Node) bool {
```
Comment (zz-jason) [inline executor/read_only_checker.go]: maybe a better way is put it into package `ast` or `plan` ?

File: executor/prepared.go [hunk 4]
```diff
@@ -287,6 +287,7 @@ func CompileExecutePreparedStmt(ctx context.Context, ID uint32, args ...interfac
 	stmt := &ExecStmt{
 		InfoSchema: GetInfoSchema(ctx),
 		Plan:       execPlan,
+		ReadOnly:   IsReadOnly(execStmt),
```
Comment (coocood) [inline executor/prepared.go]: This check `ast.ExecuteStmt` which is always false.

File: executor/read_only_checker.go (hunk 3 above)
Comment (wentaoxu) [reply executor/read_only_checker.go]: Yes, I think package ```ast``` is a better place

**Comment by wentaoxu** [comment]
PTAL @coocood @shenli @zz-jason

**Comment by coocood** [comment]
LGTM

**Pull request merged**